### PR TITLE
Backend feature portfolio

### LINKED ISSUE #278: Backend Add Portfolio Feature
As we indicated in requirements analysis:

1.1.2.6.1 User shall have at least one portfolio
1.1.2.6.2 User shall be able to have different portfolios
1.1.2.6.3 User shall be able to rename portfolio
1.1.2.6.4 Users shall be able to add or remove trading equipments from/to portfolio.
1.1.2.6.5 User shall be able to share portfolio in profile page
1.1.2.6.6 Other users shall be able to follow other user's portfolio.
1.1.2.6.7 Users shall be able to create/delete their portfolios.

### PR DESCRIPTION
**Pull request by cemalaytekin**
closes #278

### CODE REVIEW

File: app/backend/controllers/portfolio.js [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+/*
+  Patch method for editing portfolios.
+*/
+module.exports.renamePortfolio = async (request, response) => {
+  let Portfolio = request.models['Portfolio']
+  const userId = request.session['user']._id
+  const title = request.body["title"];
+  const definition = request.body["definition"];
+  const PortfolioId = request.body['PortfolioId'];
```
Comment (gurkandemir) [inline app/backend/controllers/portfolio.js]: it would be better if we get id from url

File: app/backend/controllers/portfolio.js [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+  /*
+  Post method for portfolio.
+  It saves portfolio to database.
+*/
+module.exports.addTradingEq = async (request, response) => {
+  let Portfolio = request.models['Portfolio']
+  let PortfolioTradingEq = request.models['PortfolioTradingEq']
+  let tradingEq = request.body['tradingEq']
+  const PortfolioId = request.body['PortfolioId'];
```
Comment (gurkandemir) [inline app/backend/controllers/portfolio.js]: same

File: app/backend/controllers/portfolio.js [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+  // portfolioTradingEq instance to add to the database
+  let portfolioTradingEq = new PortfolioTradingEq({
```
Comment (gurkandemir) [inline app/backend/controllers/portfolio.js]: what if equipment already exists?

File: app/backend/controllers/portfolio.js [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+
+/*
+  Delete method for remove trading eq. from portfolio.
+  It deletes given portfolio - trading eq. instances.
+*/
+module.exports.removeTradingEq = async (request, response) => {
+  let Portfolio = request.models['Portfolio']
+  let PortfolioTradingEq = request.models['PortfolioTradingEq']
+  let tradingEq = request.body['tradingEq']
+  const PortfolioId = request.body['PortfolioId'];
```
Comment (gurkandemir) [inline app/backend/controllers/portfolio.js]: same

File: app/backend/controllers/portfolio.js [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+      });
+  } else {
+    return response.status(400).send({
+      errmsg: "No such portfolio."
+    })
+  }
+}
+
+/*
+  Post method for following specific trading equipment.
```
Comment (gurkandemir) [inline app/backend/controllers/portfolio.js]: wrong comment. trading equipment?

**Comment by gurkandemir** [comment]
we keep portfolio's information about its publicity but we did not use this attribute in any endpoints. in my opinion users can see other's portfolios if they are public.

File: app/backend/controllers/profile.js [hunk 6]
```diff
@@ -7,12 +7,14 @@ const {findUserFollows, checkPassword, checkIBAN, checkTCKN} = require('../utils
...
 */
 
-async function profileResponse(user, me, followStatus, TradingEqFollow, Article) {
+async function profileResponse(user, me, followStatus, TradingEqFollow, Article, Portfolio) {
   let followings = await findUserFollows({ FollowingId: user._id, status: true })
   let followers = await findUserFollows({ FollowedId: user._id, status: true })
   let followRequests = await findUserFollows({ FollowedId: user._id, status: false})
   let followingTradings = await TradingEqFollow.find({ UserId : user._id })
   let articles = await Article.find().where('userId').equals(user._id)
+  let portfolios = await Portfolio.find().where('userId').equals(user._id)
```
Comment (gurkandemir) [inline app/backend/controllers/profile.js]: we can also check public control here, but if I am looking my own portfolios I have to see all portfolios no matter it is public or not.

File: app/backend/controllers/portfolio.js [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+                  TradingEq: tradingEq
+              },
+              PortfolioId: doc._id,
+              TradingEq: tradingEq
+          });
+
+          // Saves the instance into the database, returns any error occured
+          portfolioTradingEq.save()
+      }
+      return response.status(204).send();
```
Comment (bahadirh) [inline app/backend/controllers/portfolio.js]: We might want to return the newly-created portfolio id to the user.

File: app/backend/controllers/portfolio.js (hunk 3 above)
Comment (cemalaytekin) [reply app/backend/controllers/portfolio.js]: return database duplication error, since portfolio id and trading eq form _id which is unique.

**Pull request merged**